### PR TITLE
[BNNS]Fix the issue for ADD/MUL/CONCATENATION

### PR DESCRIPTION
**Pull request by lisa0314**
Fix [intel/webml-polyfill#732](https://github.com/intel/webml-polyfill/issues/732). 
- cts-test: run normally
- squeenet: run well
- mobilenet v1: run well in hybrid mode

### CODE REVIEW

File: services/ml/compilation_impl_mac.mm [hunk 1]
```diff
@@ -61,8 +61,10 @@
              << "preference: " << preference;
   if (@available(macOS 10.13, *)) {
     if ((is_bnns_ = (preference == mojom::PREFER_FAST_SINGLE_ANSWER))) {
+      LOG(ERROR) << "CompileModelWithBNNS";
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac.mm]: Unnecessary.

File: services/ml/compilation_impl_mac.mm [hunk 2]
```diff
@@ -61,8 +61,10 @@
...
       CompileModelWithBNNS(std::move(callback));
     } else {
+      LOG(ERROR) << "CompileModelWithMPS";
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac.mm]: Ditto.

File: services/ml/compilation_impl_mac_bnns.mm [hunk 3]
```diff
@@ -65,7 +65,7 @@ void ComputeBNNSOffsetForImplicitPadding(bool same_padding,
 
 API_AVAILABLE(macosx(10.13))
 bool CompileResizeBilinearBNNS(OperationMac& operation) {
-  DLOG(INFO) << "CompilationImplMac::CompileResizeBilinearBNNS";
+  LOG(ERROR) << "CompilationImplMac::CompileResizeBilinearBNNS";
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac_bnns.mm]: DLOG(INFO)

File: services/ml/compilation_impl_mac_bnns.mm [hunk 4]
```diff
@@ -80,7 +80,7 @@ bool CompileArithmeticBNNS(OperationMac& operation,
                            const std::map<uint32_t, ValueInfo>& values,
                            const std::unique_ptr<int8_t[]>& memory,
                            const std::vector<OperandMac>& operands) {
-  DLOG(INFO) << "CompilationImplMac::CompileArithmetic";
+  LOG(ERROR) << "CompilationImplMac::CompileArithmetic";
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac_bnns.mm]: Ditto.

File: services/ml/compilation_impl_mac_bnns.mm [hunk 5]
```diff
@@ -97,6 +97,46 @@ bool CompileArithmeticBNNS(OperationMac& operation,
     return false;
   }
 
+  const OperandMac& input = operands[inputs[0]];
+  for (size_t i = 0; i < 2; i++) {
+    uint32_t extend_input_idx = inputs[i];
+    if (values.find(extend_input_idx) != values.end()) {
+      LOG(ERROR) << "extend_input_idx: " << extend_input_idx;
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac_bnns.mm]: Ditto.

File: services/ml/compilation_impl_mac_bnns.mm [hunk 6]
```diff
@@ -97,6 +97,46 @@ bool CompileArithmeticBNNS(OperationMac& operation,
...
+          memory.get() + values.at(extend_input_idx).offset);
+      if (input.dimensions.size() == 4) {
+        const int32_t input_batch = input.dimensions[0];
+        const int32_t input_height = input.dimensions[1];
+        const int32_t input_width = input.dimensions[2];
+        const int32_t input_depth = input.dimensions[3];
+        LOG(ERROR) << "INPUT_WIDTH: " << input_width;
+        LOG(ERROR) << "INPUT_HEIGHT: " << input_height;
+        LOG(ERROR) << "INPUT_BATCH: " << input_batch;
+        LOG(ERROR) << "INPUT_DEPTH: " << input_depth;
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac_bnns.mm]: Ditto.

File: services/ml/compilation_impl_mac_bnns.mm [hunk 7]
```diff
@@ -142,7 +182,7 @@ bool CompileConv2DBNNS(OperationMac& operation,
                        const std::map<uint32_t, ValueInfo>& values,
                        const std::unique_ptr<int8_t[]>& memory,
                        const std::vector<OperandMac>& operands) {
-  DLOG(INFO) << "CompilationImplMac::CompileConv2D";
+  LOG(ERROR) << "CompilationImplMac::CompileConv2D";
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac_bnns.mm]: DLOG(INFO)

File: services/ml/compilation_impl_mac_bnns.mm [hunk 8]
```diff
@@ -307,7 +347,7 @@ bool CompileAverageOrMaxPool2DBNNS(OperationMac& operation,
                                    const std::map<uint32_t, ValueInfo>& values,
                                    const std::unique_ptr<int8_t[]>& memory,
                                    const std::vector<OperandMac>& operands) {
-  DLOG(INFO) << "CompilationImplMac::CompileAverageOrMaxPool2DBnns";
+  LOG(ERROR) << "CompilationImplMac::CompileAverageOrMaxPool2DBnns";
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac_bnns.mm]: Ditto.

File: services/ml/compilation_impl_mac_bnns.mm [hunk 9]
```diff
@@ -538,8 +578,44 @@ bool CompileConcatenationBNNS(OperationMac& concat,
   std::vector<uint32_t> inputs = concat.inputs;
   std::vector<uint32_t> outputs = concat.outputs;
 
+  for (size_t i = 0; i < inputs.size() - 1; ++i) {
+    uint32_t extend_input_idx = inputs[i];
+    if (values.find(extend_input_idx) != values.end()) {
+      float* input_memory =
+          reinterpret_cast<float*>(memory.get() + values.at(inputs[i]).offset);
+      const OperandMac& input = operands[inputs[i]];
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac_bnns.mm]: Can you extract these section to one function with https://github.com/otcshare/chromium-src/pull/119/commits/26f5bfc088459e15c53cc9327b2d478eeb1b6c6d

File: services/ml/execution_impl_mac_bnns.mm [hunk 10]
```diff
@@ -145,19 +145,20 @@ bool TransposeForInput(const ml::OperationMac& operation,
...
-                  operation_input_idx = input_idx;
-                  operation_input = compilation_->operands_[input_idx];
+                  // operation_input_idx = input_idx;
+                  operation_input = compilation_->operands_[operation_input_idx];
                   input_flag = j;
                   break;
                 }
               }
-
+              LOG(ERROR) << "i: " << i;
```
Comment (fujunwei) [inline services/ml/execution_impl_mac_bnns.mm]: Unnecessary.

File: services/ml/execution_impl_mac_bnns.mm [hunk 11]
```diff
@@ -184,10 +187,22 @@ bool TransposeForInput(const ml::OperationMac& operation,
                   src[i] = raw_input;
                 }
                 is_outer_input = false;
-              } else {
-                if (operation.local_operation == KAdd ||
-                    operation.local_operation == KMul ||
-                    operation.local_operation == KConcatenation || i == 0) {
+              } else if (i == 0) {
+                LOG(ERROR) << "src[" << i << "] = bnns_operands_memory_map_[" << operation_input_idx << "]";
```
Comment (fujunwei) [inline services/ml/execution_impl_mac_bnns.mm]: DLOG(INFO)

File: services/ml/ml_service.cc [hunk 12]
```diff
@@ -22,7 +22,22 @@ MLService::MLService(service_manager::mojom::ServiceRequest request)
 MLService::~MLService() = default;
 
 void MLService::OnStart() {
-#if defined(OS_LINUX) || defined(OS_WIN)
+#if defined(OS_MACOSX)
```
Comment (fujunwei) [inline services/ml/ml_service.cc]: No relative code.

File: services/ml/compilation_impl_mac.mm (hunk 1 above)
Comment (lisa0314) [reply services/ml/compilation_impl_mac.mm]: Sorry for putting the experimental commit up. I will update soon.

**Comment by lisa0314** [comment]
@fujunwei , Please help to review again.

File: services/ml/compilation_impl_mac.mm (hunk 2 above)
Comment (fujunwei) [reply services/ml/compilation_impl_mac.mm]: Remove the output.

File: services/ml/compilation_impl_mac_bnns.mm [hunk 13]
```diff
@@ -552,7 +606,7 @@ bool CompileFullyConnectedBNNS(OperationMac& operation,
                                const std::map<uint32_t, ValueInfo>& values,
                                const std::unique_ptr<int8_t[]>& memory,
                                const std::vector<OperandMac>& operands) {
-  DLOG(INFO) << "CompilationImplMac::CompileFullyConnectedBNNS";
+  LOG(ERROR) << "CompilationImplMac::CompileFullyConnectedBNNS";
```
Comment (fujunwei) [inline services/ml/compilation_impl_mac_bnns.mm]: DLOG(INFO)

**Comment by lisa0314** [comment]
@fujunwei , Please help to review again.

File: services/ml/ml_utils_mac.h [hunk 14]
```diff
@@ -50,6 +50,7 @@ struct OperationMac : public Operation {
   int input_batch_size;
   uint32_t offset_x;
   uint32_t offset_y;
+  std::vector<float*> extend_input;
```
Comment (fujunwei) [inline services/ml/ml_utils_mac.h]: The data need to be freed.

Comment (lisa0314) [reply services/ml/ml_utils_mac.h]: These data will be set to `src` and `src` will bre freed.

Comment (fujunwei) [reply services/ml/ml_utils_mac.h]: Get it, thanks.

**Comment by fujunwei** [comment]
LGTM.

**Pull request merged**